python/ray/data/tests/test_execution_optimizer.py

```diff
@@ -1698,6 +1698,30 @@ def test_zero_copy_fusion_eliminate_build_output_blocks(ray_start_regular_shared
     )
 
 
+def test_logical_optimization_e2e():
+    ds = ray.data.range(10, override_num_blocks=2)
+    ds = ds.randomize_block_order()
+    ds = ds.map_batches(lambda x: x)
+    ds = ds.randomize_block_order()
+    assert set(extract_values("id", ds.take_all())) == set(range(10))
+
+    # Expected execution plan:
+    # InputDataBuffer[Input]
+    # -> TaskPoolMapOperator[ReadRange->MapBatches(<lambda>)]
+    # -> AllToAllOperator[RandomizeBlockOrder]
```

Comment:
Please assert on the shape of the whole plan

Reply:
What's the most elegant way to do so ? don't want to be inspecting too many private methods that are likely to change

Reply:
FWICT I am checking the shape implicitly by looking for asserting "Operator 2" is ds.stats and that "Operator 3" is not

Reply:
Just stringize whole plan and assert on its output